File: 101-200/152. Maximum Product Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int up = nums[0], down = nums[0], length = nums.length, res = nums[0];
        for (int i = 1; i < length; i++){
            int up_pre = up, down_pre = down;
            if(nums[i]>0){
                up = Math.max(up_pre*nums[i],nums[i]);
                down = Math.min(down_pre*nums[i],nums[i]);
            }
            else{
                up = Math.max(down_pre*nums[i],nums[i]);
                down = Math.min(up_pre*nums[i],nums[i]);
            }
            res = Math.max(res, up);
        }
        return res;
    }
}

/*
Topic: Dynamice programming (DP) and array
TC: O(n) SC: O(1)
Method: iterate array and use 'up' and 'down' to record max and min
product among the subarrays whose last element is nums[i]. Compare up with res
every iteration. Return res eventually.
*/

File: 101-200/152. Maximum Product Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        l = len(nums)
        up = down = nums[0]
        res= nums[0]

        for i in range(1,l):
            up_pre = up
            down_pre = down
            up = max(up_pre*nums[i],nums[i],down_pre*nums[i])
            down = min(up_pre*nums[i],nums[i],down_pre*nums[i])
            res = max(res,up)
        return res

"""
Topic: Dynamice programming (DP) and array
TC: O(n) SC: O(1)
Method: iterate array and use 'up' and 'down' to record max and min
product among the subarrays whose last element is nums[i]. Compare up with res
every iteration. Return res eventually.
"""
